design_docs/protostar-architecture.md

```diff
@@ -0,0 +1,63 @@
+# Protostar modular architecture
+
+This document proposes the idea how to modularize Protostar. Goals of the modularization are
+- Allow to introduce better ownerships of the codebase parts
+- Make workflow asynchronous between smaller sub-teams operating on contracts between modules
+- Allow incremental rewrite to Rust
+- Remove compilation responsibility from Protostar codebase
+
+
+
+## General overview
+First decomposition of Protostar will introduce 3 modules
+
+- Protostar Gateway
+- Protostar Starknet
+- Protostar Testing backend
+
+Each of the modules would be a separate poetry package (open question?), to simplify scenarios of updating different dependencies. Example of such scenario, Protostar-starknet relies on starknet.py and devnet but testing backend does not. As a final goal everything would be under one cargo workspace.
```

Comment:
Some of the modules will be crates though (protostar gateway), right? What are the benefits of separating those into separate packages as of now?

Reply:
>Example of such scenario, Protostar-starknet relies on starknet.py and devnet but testing backend does not.
This was a scenario, easier version upgrades

Reply:
I think what would be beneficial is to split Protostar into two apps: test and starknet

Reply:
And if so - is it only for us (developers) to make the development easier or do we want to communicate that fact to end users?

Reply:
@piotmag769 Undelrying architecture is for us, but we are going to communicate with users that protostar will be accesible through scarb